File: lib/models/availability.dart
// Models for `csdo_api/availability.php` — how much of each asset is free
// for a specific loan window, so the approval picker and the new-request
// form can show real "N free for these dates" figures instead of a
// point-in-time "is it out right now" guess.

/// One approved request that already has a claim on an asset for a window
/// overlapping the one being checked — surfaced so the picker (and the
/// approval error message) can name what the clash is with.
class WindowConflict {
  const WindowConflict({
    required this.requestId,
    required this.title,
    required this.borrowOn,
    required this.returnOn,
    required this.quantity,
  });

  final int requestId;
  final String title;
  final String borrowOn;
  final String returnOn;
  final int quantity;

  /// "2026-09-15" for a same-day loan, "2026-09-15 – 2026-09-16" otherwise.
  String get rangeLabel {
    if (borrowOn.isEmpty && returnOn.isEmpty) return '';
    if (borrowOn == returnOn || returnOn.isEmpty) return borrowOn;
    if (borrowOn.isEmpty) return returnOn;
    return '$borrowOn – $returnOn';
  }

  factory WindowConflict.fromJson(Map<String, dynamic> json) => WindowConflict(
        requestId: (json['request_id'] as num?)?.toInt() ?? 0,
        title: (json['title'] as String?) ?? 'Request',
        borrowOn: (json['borrow_on'] as String?) ?? '',
        returnOn: (json['return_on'] as String?) ?? '',
        quantity: (json['quantity'] as num?)?.toInt() ?? 1,
      );
}

/// How much of one asset is free for a specific loan window.
///
/// [windowFree] is in units — 0 or 1 for an individual asset, 0..owned for a
/// bulk pool. [availableNow] is the point-in-time figure the current borrow
/// flow uses, kept so the picker can stay at least as strict as it was.
class AssetWindowAvailability {
  const AssetWindowAvailability({
    required this.tagId,
    required this.isBulk,
    required this.windowCommitted,
    required this.windowFree,
    required this.availableNow,
    this.conflicts = const [],
  });

  final String tagId;
  final bool isBulk;

  /// Units already committed to other approved requests overlapping the
  /// window.
  final int windowCommitted;

  /// Units free to lend for the whole window.
  final int windowFree;

  /// Units free right now (owned − out − damaged for a pool; 1 iff
  /// `available` for an individual asset).
  final int availableNow;

  final List<WindowConflict> conflicts;

  factory AssetWindowAvailability.fromJson(Map<String, dynamic> json) =>
      AssetWindowAvailability(
        tagId: json['tag_id'] as String,
        isBulk: (json['tracking'] as String?) == 'bulk',
        windowCommitted: (json['window_committed'] as num?)?.toInt() ?? 0,
        windowFree: (json['window_free'] as num?)?.toInt() ?? 0,
        availableNow: (json['available_now'] as num?)?.toInt() ?? 0,
        conflicts: (json['conflicts'] as List<dynamic>? ?? [])
            .map((c) => WindowConflict.fromJson((c as Map).cast<String, dynamic>()))
            .toList(),
      );
}

/// The full `availability.php` response: the resolved window plus a lookup
/// of tag ID → availability for it.
class WindowAvailabilityReport {
  const WindowAvailabilityReport({
    required this.from,
    required this.to,
    required this.byTagId,
  });

  final String from;
  final String to;
  final Map<String, AssetWindowAvailability> byTagId;

  AssetWindowAvailability? forTag(String tagId) => byTagId[tagId];

  factory WindowAvailabilityReport.fromJson(Map<String, dynamic> json) {
    final assets = (json['assets'] as List<dynamic>? ?? [])
        .map((a) =>
            AssetWindowAvailability.fromJson((a as Map).cast<String, dynamic>()))
        .toList();
    return WindowAvailabilityReport(
      from: (json['from'] as String?) ?? '',
      to: (json['to'] as String?) ?? '',
      byTagId: {for (final a in assets) a.tagId: a},
    );
  }
}

File: csdo_api/availability.php
<?php
require __DIR__ . '/db.php';

// Read-only: how much of each asset is free for a given loan window, for
// the ADMIN approval picker (lib/widgets/asset_assignment_sheet.dart). The
// response itemises the inventory (tag IDs, per-asset counts, the titles of
// clashing requests), so it must not be exposed to requesters — the
// new-request form uses request_feasibility.php, which returns only a
// coarse outlook. When this project grows real roles, gate this endpoint
// behind an admin check.
//
//   GET availability.php?from=YYYY-MM-DD&to=YYYY-MM-DD[&exclude_request=ID]
//
// `exclude_request` is the request being (re)approved — its own current
// assignment is left out of the "already committed" totals so re-opening the
// picker for an approved request doesn't count it against itself.

if ($_SERVER['REQUEST_METHOD'] !== 'GET') {
    fail(405, 'Method not allowed');
}

$from = trim($_GET['from'] ?? '');
$to = trim($_GET['to'] ?? '');
$excludeRequest = isset($_GET['exclude_request']) ? (int) $_GET['exclude_request'] : 0;

$fromIso = iso_date_or_null($from);
$toIso = iso_date_or_null($to);
if ($fromIso === null || $toIso === null) {
    fail(400, 'from and to are required and must be valid dates (YYYY-MM-DD).');
}
if ($fromIso > $toIso) {
    fail(400, 'from must be on or before to.');
}

$commitments = overlapping_asset_commitments($mysqli, $fromIso, $toIso, $excludeRequest);

$result = $mysqli->query(
    'SELECT a.id, a.tag_id, a.name, a.status, a.tracking, ' .
    'a.quantity_total, a.quantity_out, a.quantity_damaged ' .
    'FROM assets a ORDER BY a.id DESC'
);

$assets = [];
while ($row = $result->fetch_assoc()) {
    $id = (int) $row['id'];
    $isBulk = ($row['tracking'] ?? 'individual') === 'bulk';
    $entry = $commitments[$id] ?? ['committed' => 0, 'conflicts' => []];
    $committed = (int) $entry['committed'];

    if ($isBulk) {
        $owned = (int) $row['quantity_total'];
        $damaged = (int) $row['quantity_damaged'];
        // Free for the window: what's owned, less what's set aside damaged,
        // less what's committed to other approved requests that overlap it.
        $windowFree = max(0, $owned - $damaged - $committed);
        // Free right now: the point-in-time figure the current flow uses.
        $availableNow = max(0, $owned - $damaged - (int) $row['quantity_out']);
    } else {
        // An individual asset is either wholly free for the window or not.
        // It must be part of the borrowable pool ('available' or 'in_use';
        // 'maintenance' / 'in_stock' are filed out) and unclaimed by any
        // overlapping approved request.
        $lendable = in_array($row['status'], ['available', 'in_use'], true);
        $windowFree = ($lendable && $committed === 0) ? 1 : 0;
        $availableNow = $row['status'] === 'available' ? 1 : 0;
    }

    $assets[] = [
        'tag_id' => $row['tag_id'],
        'tracking' => $isBulk ? 'bulk' : 'individual',
        'window_committed' => $committed,
        'window_free' => $windowFree,
        'available_now' => $availableNow,
        'conflicts' => array_values($entry['conflicts']),
    ];
}

echo json_encode(['from' => $fromIso, 'to' => $toIso, 'assets' => $assets]);
